### DISCUSSION TITLE
Calling Ruby Method from Crystal

### DISCUSSION THREAD

**Discussion by mgomes** [Anyolite/anyolite#12] [Q&A] [answered]
Hello,

I'd like to call a method from Crystal in my Ruby code. In the test suite I see that being done with code like this:

```crystal
def class_call_test
    puts Anyolite.call_rb_method_of_class(self.class, :class_method_in_ruby, ["World", 4], cast_to: String)
end
```

The problem I am having is that `Anyolite.call_rb_method_of_class` only takes a class as the first argument. That means that I'd have to know the class name in advance and also have the same class declared in Crystal. Is there another way to do it?

Thanks!

**Comment by Hadeweka** [comment, marked as answer]
Hi,

as far as I understand this, you want to call a Ruby class method where the name of the class and the name of the method are known, but the Ruby class has no analogue in Crystal (please correct me otherwise)?

It seems that I forgot to actually implement a proper method for this case, so definitely thank you for pointing it out!
I will try to fix this in the next days.

Until then, you can do something like this:
```crystal
class_name = "TestModule::Test"
method_name = "class_method_in_ruby"
args = ["World", 4]

rb_class = rb.execute_script_line(class_name)
result = Anyolite.call_rb_method_of_object(rb_class, method_name, args, cast_to: String)
```
You get the Ruby class by just putting its name into a script line, which then returns a `RbValue` (the internal representation of the Ruby object). Then, you can just call the method name with the `RbValue` as the object.

This does not only work for classes, but all other objects which are somehow accessable from a global context.

I hope I could help you, otherwise please feel free to ask.

**Reply by mgomes** [reply]
@Hadeweka That's the exact the use case I am going for. Thanks! I'm implementing a system where some events happen in Crystal and Crystal calls out to user-defined scripts written in MRuby and so the entry point is always known.

This workaround is perfect. Thanks for considering adding native support for it as well!

**Reply by Hadeweka** [reply]
With the 0.14.0 release, the following code will be possible:

```crystal
rb_module = Anyolite::RbModule.get_from_ruby_name(rb, "TestModule")

if rb_module
  rb_class = Anyolite::RbClass.get_from_ruby_name(rb, "Test", under: rb_module)

  if rb_class
    method_name = "class_method_in_ruby"
    args = ["World", 4]

    class_obj = rb_class.to_rb_obj

    result = Anyolite.call_rb_method_of_object(class_obj, method_name, args, cast_to: String)
  end
end
```

This does on the first glance not look like an improvement, but you can actually now obtain classes and modules by their name using `Anyolite::RbClass.get_from_ruby_name` without calling any script lines. This gives a `Anyolite::RbClass` object, which in turn can be transformed into a Ruby object by calling `to_rb_obj`.

The tests are necessary because a class may not exist, in which case `rb_class` will become `nil`. In my opinion it is more in the spirit of Crystal to do so and produces safer code.

Overall, the way this can be done currently is still a bit chaotic, but I have this on my list for a future release. That's also the reason why the version number is only 0.14.0 instead of 1.0.0 ;-)

(Edit: Typo)

**Reply by mgomes** [reply]
Amazing! This will work great.

Thanks so much for the release.
